### PR TITLE
Add money factory

### PR DESCRIPTION
**Pull request by sagikazarmark**
Fixes #415

### CODE REVIEW

**Comment by willemstuursma** [comment]
Looks great, good naming too.

Only thing I can add is that I think the list of `@method` declarations should be sorted.

**Comment by willemstuursma** [comment]
That's not possible or needed, since the generated file is part of this repository anyway.

**Comment by sagikazarmark** [comment]
@willemstuursma Great idea, will do.

@Gummibeer users of this library are not supposed to run the generator script. If you need that kind of functionality you either need to write your own factory or live with the IDE not recognizing the currency.

**Comment by Gummibeer** [comment]
@sagikazarmark that's bad - my PR supported custom Currencies which ist very handy if create custom ones.

**Comment by sagikazarmark** [comment]
Again, this file is part of this library, any modification to it should be considered a serious anti-pattern and it's behavior is undefined, not supported.

You are free to do that though, but I think the solution I provided above (creating your own factory) is the best way to achieve the same functionality.

**Comment by frederikbosch** [comment]
I dont know if he likes it, but I am going to mention @Haehnchen in this conversation. He is the one who is responsible for creating the Jetbrains Symfony plugin. Maybe he knows a better method for creating type completion for our dynamic method problem.

p.s. Sorry Daniel if you don't like the @ mention.

**Comment by sagikazarmark** [comment]
I already tried it, I guess people could live with that.

But: I'm not a huge fan of this whole static factory thing, especially of the callStatic method. If we ever decide to rather generate static factory methods...with this solution we can.

Furthermore, I believe the Money class becomes a bit cleaner without the whole static factory thing, that's why I moved the callstatic to the generated trait as well.

The alternative would be writing a PHPStorm plugin, which would probably address all your concerns. I looked into it, did some research, and I needed half a day cooldown period after that, because my mind was blown.

To sum up: I think this is an acceptable compromise, even makes things a bit cleaner in terms of static hack moved out of money. It makes a huge step forward in DX, which at least at some level should be our goal too.

**Comment by frederikbosch** [comment]
> Furthermore, I believe the Money class becomes a bit cleaner without the whole static factory thing, that's why I moved the callstatic to the generated trait as well.

I agree.

> To sum up: I think this is an acceptable compromise, even makes things a bit cleaner in terms of static hack moved out of money. It makes a huge step forward in DX, which at least at some level should be our goal too.

I don't agree there. I feel our research was not broad enough. There must be more possibilities.

**Comment by frederikbosch** [comment]
Isn't there something we can do with this: https://confluence.jetbrains.com/display/PhpStorm/PhpStorm+Advanced+Metadata?

**Comment by sagikazarmark** [comment]
AFAIK it cannot add new methods, only the old API which is now deprecated.

**Comment by sagikazarmark** [comment]
I'm okay with saying we should keep looking to see if there is a better solution, but I think rejecting this solution without a viable alternative is not the way forward.

**Comment by frederikbosch** [comment]
I am not rejecting without a viable alternative. I gave one. Since we both agree on _the support for the whole static factory thing, especially of the callStatic method_, I rather feel more for dropping support for the whole `__callStatic` than extending it. At least that is a discussion we never had.

**Comment by frederikbosch** [comment]
I rather add this `Money::fromString('EUR1000')` and `(string)$money; // EUR1000`, which also gives convenience and is a similar factory, than we go forward on `__callStatic`.

**Comment by frederikbosch** [comment]
And/or something like `Money::amount('EUR', 1000);`. Then we would be rid of the `__callStatic` ugliness and the IDE completion.

**Comment by sagikazarmark** [comment]
Although it's not a nice solution, I believe it improves DX. I guess I will have to take my chances with brain damage and look at phpstorm plugins again. 😄 

https://www.jetbrains.org/intellij/sdk/docs/phpstorm/php_open_api.html

This looks exactly what we need

**Comment by sagikazarmark** [comment]
The reason why `Money::EUR` is great is because it's domain specific. Any change to this will be less domain specific. The callStatic method is not a nice way of doing anything, but in this case I don't see any other option.

I spent another hour staring at Java code and I believe besides auto-completion, we need to cheat PHPStorm to handle the references well.

The more I think about this, the more I'm willing to just accept this solution as is. I believe thinking of it as a "change for the sake of IDEs" is a limited point of view. The change is for the sake of developer experience and in this case it doesn't hurt any design principles (like all the criticism we get because of the final classes). It is tested, so if something goes wrong, we will know about it. And the code is literally the same.

The solution you proposed @frederikbosch IMO is not a good one, because it doesn't make DX any better. The IDE will just complain about a different thing.

The good thing about this solution is that any time in the future if we find a better one we can just remove this madness without breaking any API. But currently I'm not convinced if that's even possible or worth the effort. (Please take into consideration that I've just spent an hour looking at java code, which probably affected my mood in a negative way 😄 ) Also, compared to a PHPstorm specific solution, (I believe) this one is universal, so it will work for those who are stuck with something else.

**Comment by frederikbosch** [comment]
Aight. Go for it!

**Comment by sagikazarmark** [comment]
Did my arguments really convince you or you just tired of the debate? 😄

**Comment by frederikbosch** [comment]
Well, I understand that this solution is OK for now and does not bring any possible problems for the future if we want to change something. That said, I’d rather have no responsibility for IDEs and DX at all.

**Comment by sagikazarmark** [comment]
I understand. We can revisit this in the future, for now this is a good enough solution. Thanks for working towards quality. 👍

**Comment by frederikbosch** [comment]
Thank you, for your recent massive contributions to the package!

**Comment by willemstuursma** [comment]
We're very happy at Mollie with this addition!
